### PR TITLE
fix(mcp): drop unresolvable rel_paths from list_files (#176)

### LINKED ISSUE #176: list_files emits malformed rel_path for filenames containing `--` (causes open_file FILE_NOT_FOUND)
## Summary

For source files whose name contains a double-dash (`--`), `dir2mcp_list_files` returns a malformed `rel_path` with a phantom `-/` prefix and the wrong extension. Subsequent `dir2mcp_open_file` calls against that rel_path fail with `FILE_NOT_FOUND`.

## Reproduction

Corpus: BVI legal corpus (stas-legal). One file has a `--` in its name:

```
$ ls /Users/adeliev/Development/stas-legal | grep AML | grep 2025
SINo3of2025--AML(Amendment)Regulations2025.pdf
```

Call `list_files`:

```jsonc
// dir2mcp_list_files {"limit": 1}
{
  "rel_path": "-/SINo3of2025--AML(Amendment)Regulations2025.md",
  "doc_type": "md",
  "size_bytes": 3140,
  ...
}
```

Two things wrong with that `rel_path`:
1. Phantom `-/` directory prefix — there is no `-/` directory on disk.
2. `.md` extension — the source file is `.pdf`. The OCR cache lives at `.dir2mcp/cache/ocr/<sha>.md`, but that's an internal artifact and shouldn't surface as the file's external rel_path.

Round-trip:

```jsonc
// dir2mcp_open_file {"rel_path": "-/SINo3of2025--AML(Amendment)Regulations2025.md"}
{
  "structuredContent": {
    "error": {"code": "FILE_NOT_FOUND", "message": "file not found", "retryable": false}
  },
  "isError": true
}
```

In Claude Desktop this surfaces as `Failed to call tool "dir2mcp_open_file"` — which is how this was found.

## Scope

Out of 91 files in the test corpus, 1 has the bug. All other rel_paths round-trip correctly. The common factor on the affected file is the `--` (with parens, but other parens-containing names round-trip fine — so it's the double-dash that's load-bearing).

## Likely cause

Some path normalization step appears to split the filename on `--` and treat the leading segment as a directory:
- `SINo3of2025--AML...pdf` → split → `["SINo3of2025-", "AML...pdf"]`
- The first segment is then trimmed/transformed and the file is registered as `<segment>/<rest>.md` against the OCR cache.

Worth a unit test that walks a synthetic corpus with adversarial filenames (`--`, leading `-`, parens, spaces, unicode) and asserts `list_files` rel_paths round-trip through `open_file`.

## Impact

Any agent (Claude, Cursor, etc.) that drives the toolchain by calling `list_files` and then `open_file` on the first/named result will fail unpredictably for any file with `--` in its name. This is a non-trivial subset of legal/governmental corpora where multi-dash naming conventions are common.

### PR DESCRIPTION
**Pull request by yidakra**
## Summary

`dir2mcp_list_files` was returning rel_paths whose source file no longer exists on disk. Issue #176 caught one such phantom row in a real corpus (`-/SINo3of2025--AML(Amendment)Regulations2025.md`); the round-trip into `dir2mcp_open_file` then 404'd. The bug was load-bearing for any agent driving the toolchain by name.

This PR adds a resolvability gate at the `list_files` surface: filesystem-source documents are skipped when their rel_path doesn't resolve to a real file under root. Archive members (virtual paths) are exempt; if root itself can't be resolved we stand down so misconfigured deployments still get whatever the store has.

## Reproduction

In the original report:

```jsonc
// dir2mcp_list_files {"limit": 1}
{ "rel_path": "-/SINo3of2025--AML(Amendment)Regulations2025.md", "doc_type": "md", ... }

// dir2mcp_open_file {"rel_path": "-/SINo3of2025--AML(Amendment)Regulations2025.md"}
{ "isError": true, "structuredContent": { "error": { "code": "FILE_NOT_FOUND" } } }
```

The phantom rel_path is in the user's `meta.sqlite` (`source_type=filesystem`, `deleted=0`, doc_id=1) but the corresponding `-/...` directory doesn't exist on disk. It looks like leftover state from an earlier ingest path that has since been replaced; the new resolvability gate makes `list_files` safe regardless of how a stale row got there (older bug, manual edit, partial migration).

## Test plan

- [x] New regression test `TestMCPToolsCallListFiles_RoundTripsAdversarialNames` walks an adversarial corpus (`foo--bar.pdf`, `--leading.pdf`, `trail--.pdf`, parens, spaces, unicode, plain control, plus the original `SINo3of2025--AML(...)2025.pdf`), injects a stale row matching the issue report, and asserts:
  - the stale row is filtered from `list_files`
  - every adversarial real file is surfaced
  - every returned rel_path resolves to an existing file under root (the open_file precondition)
- [x] Verified the new test fails on `main` and passes with the fix applied.
- [x] Updated existing `TestMCPToolsCallListFiles_TotalReflectsHiddenFilter` and `TestMCPToolsCallListFiles_IncludeHiddenTrue` to materialise their stub docs on disk so the new gate doesn't drop them.
- [x] `make ci` passes locally (vet + cyclo + test + test-release-tools).

Closes #176.

Co-Authored-By: Claude Opus 4.7 (1M context) <noreply@anthropic.com>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved file listing to automatically exclude inaccessible or invalid files from results.
  * Enhanced handling of hidden files in directory listings.

* **Tests**
  * Added regression test for edge-case filenames and stale file exclusion.
  * Updated existing file listing tests for improved coverage.

[Review Change Stack](https://app.coderabbit.ai/change-stack/dirstral/dir2mcp/pull/181)

### CODE REVIEW

File: internal/mcp/tools.go [hunk 1]
```diff
@@ -509,6 +521,35 @@ func (s *Server) listVisibleFiles(ctx context.Context, pathPrefix, glob string,
...
+// the store has rather than an empty list.
+func (s *Server) canResolveRoot() bool {
+	rootAbs, err := filepath.Abs(strings.TrimSpace(s.cfg.RootDir))
+	if err != nil {
+		return false
+	}
+	if _, err := filepath.EvalSymlinks(rootAbs); err != nil {
+		return false
+	}
+	return true
```
[the remark this answers is not in the document]
Comment (yidakra) [reply internal/mcp/tools.go]: Addressed in 38ec543: `canResolveRoot` was already updated in 7863abf to treat empty `RootDir` as not resolvable and to verify the resolved path exists and `IsDir()`. The follow-up commit pulls the resolution into a shared `resolveRoot()` helper and adds `internal/mcp/list_files_resolve_test.go` to pin the four cases (empty, nonexistent, file-not-dir, valid directory).

File: internal/mcp/tools.go [hunk 2]
```diff
@@ -509,6 +521,35 @@ func (s *Server) listVisibleFiles(ctx context.Context, pathPrefix, glob string,
...
+// the archive itself) so they're always considered resolvable; everything else
+// must point to a real file under root.
+func (s *Server) isResolvableSource(doc model.Document) bool {
+	if strings.EqualFold(strings.TrimSpace(doc.SourceType), "archive_member") {
+		return true
+	}
+	if _, err := s.resolveDocumentPath(doc.RelPath); err != nil {
+		return false
+	}
+	return true
```
[the remark this answers is not in the document]
Comment (yidakra) [reply internal/mcp/tools.go]: Addressed in 38ec543: `listFilesFiltered` now resolves root once via `resolveRoot()` and passes the cached `rootAbs`/`rootReal` into a new package-level `isResolvableSourceWithRoot(doc, rootAbs, rootReal)` helper. The per-document hot path drops the redundant `filepath.Abs` + `EvalSymlinks(root)` calls; `resolveDocumentPath` keeps its public signature for the other call sites (`open_file`, `readDocumentFile`).

File: tests/mcp/tools_test.go [hunk 3]
```diff
@@ -1483,6 +1511,144 @@ func TestMCPToolsCallListFiles_IncludeHiddenTrue(t *testing.T) {
...
+	adversarial := []string{
+		"normal.pdf",
+		"foo--bar.pdf",
+		"--leading.pdf",
+		"trail--.pdf",
+		"with (parens).pdf",
+		"with spaces.pdf",
+		"with-单-unicode.pdf",
+		"SINo3of2025--AML(Amendment)Regulations2025.pdf",
+	}
```
[the remark this answers is not in the document]
Comment (yidakra) [reply tests/mcp/tools_test.go:1542]: Addressed in 38ec543: dropped "and a plain control" from the comment so it matches the actual corpus. Control-character filenames are tricky cross-platform and not a real corpus pattern worth defending here -- the double-dash / parens / unicode shapes already cover the issue #176 surface.
